### PR TITLE
Refactor `lists::contains`

### PR DESCRIPTION
**Pull request by ttnghia**
This is just a not-very-simple refactor to `lists::contains`:
 * Remove some wrong lines in doxygen of `lists/contains.hpp`, and rewrite doxygen there a little bit.
 * Add more comments to the code.
 * Reduce the number of code paths of the template struct functors.
 * Rename some/many variables, and reorganize code to make it cleaner.

No new feature is added in this PR, just modifying the existing functions and moving things around.

This PR is extracted from the bigger PR for easier review. The original PR is https://github.com/rapidsai/cudf/pull/10548 for supporting nested type in `lists::contains`. As such, this blocks it.

### CODE REVIEW

File: cpp/src/lists/contains.cu [hunk 1]
```diff
@@ -44,163 +41,156 @@
...
                  "Type/Scale of search key does not match list column element type.");
-    CUDF_EXPECTS(search_key.type().id() != type_id::EMPTY, "Type cannot be empty.");
+    CUDF_EXPECTS(search_keys.type().id() != type_id::EMPTY, "Type cannot be empty.");
 
     auto constexpr search_key_is_scalar = std::is_same_v<SearchKeyType, cudf::scalar>;
-
-    if constexpr (search_keys_have_nulls && search_key_is_scalar) {
+    if (search_key_is_scalar && search_keys_have_nulls) {
+      // If the scalar key is invalid/null, the entire output column will be all nulls.
       return make_numeric_column(data_type(type_id::INT32),
```
Comment (bdice) [inline cpp/src/lists/contains.cu:194]: Avoid hard-coding `INT32` when we mean `size_type`. This is the same pattern followed below.
```suggestion
      return make_numeric_column(data_type{cudf::type_to_id<size_type>()},
```

**Review by bdice — COMMENTED** [review]
Initial comments attached.

File: cpp/src/lists/contains.cu [hunk 2]
```diff
@@ -44,163 +41,156 @@
...
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_end(list_device_view const& list)
+/**
+ * @brief Return a pair of index iterators {begin, end} to loop through elements within a list.
+ *
+ * Depending on the value of `find_first`, a pair of forward or reverse iterators will be returned,
+ * allowing to loop through elements in the list by first-to-last or last-to-first order.
+ *
+ * Note that the indices are `0`-based: The first element in any list is always at `0` index.
+ *
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:67]: I think so. Conventionally, we have index `0` starting at the first row of a column. Here we have many lists in a column, the first element in any list will restart its index at index `0`.

File: cpp/src/lists/contains.cu [hunk 3]
```diff
@@ -44,163 +41,156 @@
...
-
-        auto list = cudf::list_device_view(d_lists, row_index);
-        if (list.is_null()) { return {absent_index, false}; }
-
-        auto const position = find_option == duplicate_find_option::FIND_FIRST
-                                ? finder<duplicate_find_option::FIND_FIRST>{}(list, search_key)
-                                : finder<duplicate_find_option::FIND_LAST>{}(list, search_key);
-        return {position, true};
-      });
+    CUDF_FAIL("Unsupported type in `dispatch_index_of` functor.");
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:173]: It is reachable, as we don't handle other data types such as dictionary type.

File: cpp/src/lists/contains.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:67]: Probably I need to change that comment a little bit.

File: cpp/src/lists/contains.cu [hunk 4]
```diff
@@ -44,384 +41,345 @@
...
+                   out_begin,
+                   out_begin + lists.size(),
+                   [lists = cudf::detail::lists_column_device_view{*lists_cdv_ptr}] __device__(
+                     auto const list_idx) {
+                     auto const list = list_device_view{lists, list_idx};
+                     return list.is_null() ||
+                            thrust::any_of(thrust::seq,
+                                           thrust::make_counting_iterator(0),
+                                           thrust::make_counting_iterator(list.size()),
+                                           [&list](auto const idx) { return list.is_null(idx); });
```
Comment (bdice) [inline cpp/src/lists/contains.cu:351]: Maybe you can check if the list has a null mask (is nullable) once, and then call `is_null_nocheck` in the algorithm. I'm not 100% sure if those APIs exist for `list_device_view` but I suspect they do.

File: cpp/include/cudf/lists/contains.hpp [hunk 5]
```diff
@@ -37,8 +37,6 @@ namespace lists {
  * Output `column[i]` is set to null if one or more of the following are true:
  *   1. The search key `search_key` is null
  *   2. The list row `lists[i]` is null
- *   3. The list row `lists[i]` does not contain the search key, and contains at least
```
Comment (mythrocks) [inline cpp/include/cudf/lists/contains.hpp:40]: I thought I had removed this line as part of #9510. Thanks for cleaning this up.

File: cpp/src/lists/contains.cu [hunk 6]
```diff
@@ -44,384 +41,345 @@
...
   CUDF_EXPECTS(search_keys.size() == lists.size(),
                "Number of search keys must match list column size.");
 
-  return to_contains(
-    index_of(lists, search_keys, duplicate_find_option::FIND_FIRST, stream), stream, mr);
+  return to_contains(index_of(lists,
+                              search_keys,
+                              duplicate_find_option::FIND_FIRST,
+                              stream,
+                              rmm::mr::get_current_device_resource()),
```
Comment (mythrocks) [inline cpp/src/lists/contains.cu:322]: Why was this change necessary? The `get_current_device_resource()` is implied from the defaults.
(If the intention was to also remove that default for the detail API, why?)

File: cpp/src/lists/contains.cu [hunk 7]
```diff
@@ -44,384 +41,345 @@
...
                                  rmm::cuda_stream_view stream,
                                  rmm::mr::device_memory_resource* mr)
 {
-  return to_contains(
-    index_of(lists, search_key, duplicate_find_option::FIND_FIRST, stream), stream, mr);
+  return to_contains(index_of(lists,
+                              search_key,
+                              duplicate_find_option::FIND_FIRST,
+                              stream,
+                              rmm::mr::get_current_device_resource()),
```
Comment (mythrocks) [inline cpp/src/lists/contains.cu:305]: Similar to the question below: Why be verbose about the choice of `mr`, when it's already in the defaults?

File: cpp/src/lists/contains.cu (hunk 4 above)
Comment (bdice) [reply cpp/src/lists/contains.cu:351]: This is computing the output null mask for each row by checking if the input list row is null OR the list contains a null value. Another approach is to do a bitmask AND of a segmented nullmask reduction and the input (parent) null mask, then move-assign that to the output column's null mask. I have been meaning to write this into a utility function because the pattern is fairly common. I have a draft of this logic in #10451 but it needs to be simplified and put into a common location.

File: cpp/src/lists/contains.cu [hunk 8]
```diff
@@ -44,384 +41,345 @@
...
-{
-  if (find_iter == end) {
-    return absent_index;  // Not found.
+/**
+ * @brief Functor for searching indices of the given key(s) within each list row in a list column.
+ */
+template <typename Type, typename Enable = void>
+struct search_index_fn {
+  template <typename... Args>
+  static void invoke(Args&&...)
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply cpp/src/lists/contains.cu:92]: I hadn't considered this before. This is an interesting point. I was shown the `Args&&...` technique to prevent having to repeat the signature for the "everything else" part of a dispatch endpoint. Good catch.

File: cpp/src/lists/contains.cu [hunk 9]
```diff
@@ -44,384 +41,345 @@
...
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return find_iter - begin;  // Distance of find_position from begin.
-  } else {
-    return end - find_iter - 1;  // Distance of find_position from end.
+template <typename Type>
+struct search_index_fn<Type, std::enable_if_t<is_supported_non_nested_type<Type>()>> {
+  template <typename SearchKeyIter, typename OutputPairIter>
+  static void invoke(cudf::detail::lists_column_device_view const& lists,
+                     SearchKeyIter const& keys_iter,
+                     OutputPairIter const& out_iter,
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply cpp/src/lists/contains.cu:103]: Pardon me, but what's the problem  with using `const&` in this context?

**Review by mythrocks — COMMENTED** [review]
I'm not certain that *all* the changes here are strictly necessary. (For instance, some of the variable name changes are subjective.)

But I really like the material improvements, vis-à-vis the iterator changes, combining/streamlining the utility functions, etc.

Some minor clarifications requested.

File: cpp/src/lists/contains.cu (hunk 9 above)
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:103]: If not using `const&`, the iterator will be copied, which may be more expensive than dereferencing. If we supply the iterators into a device functor then definitely we should copy it. But here we just pass them to a host function.

[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:103]: Imagine that an iterator may contain a lot of stuff inside (like `device_column_view`, which may have a vector of many `device_column_view` of children column). Passing `const&` iterator should be used in such cases.

File: cpp/src/lists/contains.cu [hunk 10]
```diff
@@ -44,384 +41,345 @@
...
+                       auto const key_opt = keys_iter[list_idx];
+                       // A null key will also result in a null output row.
+                       if (!key_opt) { return {NOT_FOUND_SENTINEL, false}; }
+
+                       auto const key = key_opt.value();
+                       return {find_option == duplicate_find_option::FIND_FIRST
+                                 ? search_list<true>(list, key)
+                                 : search_list<false>(list, key),
+                               true};
+                     });
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:126]: `key_opt` is a `thrust::optional` and the result accessed from `keys_iter` is always a value. So it doesn't matter here :smiling_face_with_tear: .

File: cpp/src/lists/contains.cu [hunk 11]
```diff
@@ -44,384 +41,345 @@
...
-      return {rmm::device_buffer{0, stream, mr}, size_type{0}};
-    } else {
-      return cudf::detail::valid_if(
-        result_validity.begin<bool>(), result_validity.end<bool>(), thrust::identity{}, stream, mr);
-    }
+template <typename SearchKeyType>
+auto get_search_keys_device_view_ptr(SearchKeyType const& search_keys,
+                                     [[maybe_unused]] rmm::cuda_stream_view stream)
+{
+  if constexpr (std::is_same_v<SearchKeyType, cudf::scalar>) {
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:157]: Yes :smiley: . This is to prepare for adding another code path in https://github.com/rapidsai/cudf/pull/10548. I add this function anyway to minimize changes in the later PR.

File: cpp/src/lists/contains.cu (hunk 4 above)
Comment (ttnghia) [reply cpp/src/lists/contains.cu:351]: Unfortunately that has not yet implemented in `list_device_view`. Maybe it is worth to have another follow up PR.

Comment (ttnghia) [reply cpp/src/lists/contains.cu:351]: Filed an FEA issue: https://github.com/rapidsai/cudf/issues/11072

File: cpp/src/lists/contains.cu [hunk 12]
```diff
@@ -44,384 +41,345 @@
...
+
+    auto out_positions = make_numeric_column(
+      data_type{type_to_id<size_type>()}, lists.size(), cudf::mask_state::UNALLOCATED, stream, mr);
+    auto out_validity   = rmm::device_uvector<bool>(lists.size(), stream);
+    auto const out_iter = thrust::make_zip_iterator(
+      out_positions->mutable_view().template begin<size_type>(), out_validity.begin());
+
+    auto const keys_iter = cudf::detail::make_optional_iterator<Type>(
+      *keys_dv_ptr, nullate::DYNAMIC{search_keys_have_nulls});
+    search_index_fn<Type>::invoke(lists_cdv, keys_iter, out_iter, find_option, stream);
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:220]: `search_index_fn` will be specialized for non-nested types and nested types. The specialization for nested types can be seen here: https://github.com/rapidsai/cudf/pull/10548/files#diff-d66d386d623323d882af299672d1a96abf1c8ade91dcf2cf0ac529e3b45fb329R140. This refactor is for adding nested types support.

File: cpp/src/lists/contains.cu [hunk 13]
```diff
@@ -44,384 +41,345 @@
...
-
-auto get_search_keys_device_iterable_view(cudf::column_view const& search_keys,
-                                          rmm::cuda_stream_view stream)
-{
-  return column_device_view::create(search_keys, stream);
-}
+/**
+ * @brief A sentinel value used for marking that a given key has not been found in the search list.
+ */
+auto constexpr __device__ NOT_FOUND_SENTINEL = size_type{-1};
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:51]: Oh no, I just learned that ALL_CAP should be used for global constants. Feel free to give your opinion on this change.

**Comment by ttnghia** [comment]
> I'm not certain that _all_ the changes here are strictly necessary. (For instance, some of the variable name changes are subjective.)

Sorry for the confusion that many changes in this PR don't sound very useful and unnecessary here. But they are necessary for the follow up PR. I want to minimize the changes in the new PR (which has new feature and new code) thus have to make more changes with the old code (without new feature implemented).

File: cpp/src/lists/contains.cu [hunk 14]
```diff
@@ -44,384 +41,345 @@
...
+};
 
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return find_iter - begin;  // Distance of find_position from begin.
-  } else {
-    return end - find_iter - 1;  // Distance of find_position from end.
+template <typename Type>
+struct search_index_fn<Type, std::enable_if_t<is_supported_non_nested_type<Type>()>> {
+  template <typename SearchKeyIter, typename OutputPairIter>
+  static void invoke(cudf::detail::lists_column_device_view const& lists,
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:101]: Sure. I prefer `operator()` too. I just follow the other currently used style: 
 * Dispatch functor used in `type_dispatcher` has `temlate<> operator()`, and
 * The dispatch functor calls another functor which has `static Type invoke()`.

File: cpp/src/lists/contains.cu (hunk 9 above)
[the remark this answers is not in the document]
Comment (bdice) [reply cpp/src/lists/contains.cu:103]: I _think_ it should be okay to take these by `const&`. My uncertainty comes from recognizing that this pattern is only used in one other place in libcudf. Do you think we should be taking most templated iterator types by `const&` in libcudf? Have we just missed that opportunity in a lot of places? Does this interfere with template type deduction somehow? cc: @jrhemstad

File: cpp/src/lists/contains.cu (hunk 10 above)
[the remark this answers is not in the document]
Comment (bdice) [reply cpp/src/lists/contains.cu:126]: But `key` can be a reference, right? We shouldn't need to construct a `thrust::optional<T>` and also ` T` containing the same value.

[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:126]: Oh yes, `key` can be a reference as `thrust::optional`. Thanks.

File: cpp/src/lists/contains.cu (hunk 14 above)
[the remark this answers is not in the document]
Comment (bdice) [reply cpp/src/lists/contains.cu:101]: This sounds like a "dispatch to invoke" pattern as in draft PR #8217.

File: cpp/src/lists/contains.cu (hunk 6 above)
Comment (ttnghia) [reply cpp/src/lists/contains.cu:322]: Hah, yes. Thanks for catching it. I reverse that.

File: cpp/src/lists/contains.cu (hunk 14 above)
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:101]: Some examples:
 * https://github.com/rapidsai/cudf/blob/4e66281f48c55735edb4b610e0f859ee2de32a75/cpp/src/groupby/sort/group_scan_util.cuh#L49
 * https://github.com/rapidsai/cudf/blob/4e66281f48c55735edb4b610e0f859ee2de32a75/cpp/src/groupby/sort/group_single_pass_reduction_util.cuh#L107
 * https://github.com/rapidsai/cudf/blob/4e66281f48c55735edb4b610e0f859ee2de32a75/cpp/src/reductions/scan/scan_inclusive.cu#L109
 * https://github.com/rapidsai/cudf/blob/4e66281f48c55735edb4b610e0f859ee2de32a75/cpp/src/lists/sequences.cu#L73

File: cpp/src/lists/contains.cu (hunk 8 above)
[the remark this answers is not in the document]
Comment (bdice) [reply cpp/src/lists/contains.cu:92]: There will be two “free” template parameters for the iterators but you can constrain it to five arguments with specific types for the 1st, 4th, and 5th parameters. Just copy the template and types from the other one.

[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:92]: I could make these `invoke` function standalone template functions instead. However, each of them depends on a separate `search_list` function so I decided to put them into different specializations of template struct.

File: cpp/src/lists/contains.cu (hunk 9 above)
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/lists/contains.cu:103]: Good catch @bdice . I noticed the iterator parameter is heavily used in `cub` and `thrust` and it seems never to be passed via const ref. Not sure why is that.

[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/src/lists/contains.cu:103]: Do not pass iterators by reference.

File: cpp/src/lists/contains.cu [hunk 15]
```diff
@@ -44,296 +41,243 @@
...
+ * @return A pair of {begin, end} iterators to iterate through the range `[0, size)` by forward or
+ *         reverse order.
+ */
+template <bool find_first>
+auto __device__ element_index_pair_iter(size_type const size)
 {
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_end<ElementType>();
+  if constexpr (find_first) {
+    return thrust::pair(thrust::make_counting_iterator<size_type>(0),
```
Comment (karthikeyann) [inline cpp/src/lists/contains.cu:77]: CTAD! Nice.

File: cpp/src/lists/contains.cu (hunk 13 above)
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/src/lists/contains.cu:51]:
```suggestion
auto constexpr NOT_FOUND_SENTINEL = size_type{-1};
```

File: cpp/src/lists/contains.cu [hunk 16]
```diff
@@ -44,296 +41,241 @@
...
-                    [] __device__(auto i) { return i != absent_index; });
+                    positions_begin + key_positions->size(),
+                    result->mutable_view().template begin<bool>(),
+                    [] __device__(auto const i) {
+                      // position == NOT_FOUND_SENTINEL: the list does not contain the search key.
+                      return i != NOT_FOUND_SENTINEL;
+                    });
+
+  auto const null_count                    = key_positions->null_count();
   [[maybe_unused]] auto [_, null_mask, __] = key_positions->release();
```
Comment (karthikeyann) [inline cpp/src/lists/contains.cu:236]: It would be good to avoid double underscore in any names.
https://en.cppreference.com/w/cpp/language/identifiers
> the identifiers with a double underscore anywhere are reserved;

**Review by karthikeyann — CHANGES_REQUESTED** [review]

File: cpp/src/lists/contains.cu [hunk 17]
```diff
@@ -44,296 +41,241 @@
...
 
 /**
- * @brief Functor to search each list row for the specified search keys.
+ * @brief Create a device pointer to the search key(s).
+ *
+ * The returned pointer will be used to construct a validity iterator or an optional iterator for
+ * the keys, which also have the same interface for both `scalar` and `column_device_view`.
+ *
+ * @return Depending on the type of the input key(s), a `scalar` pointer or a `column_device_view`
+ *         pointer will be returned.
```
Comment (jrhemstad) [inline cpp/src/lists/contains.cu:129]: This seems incredibly error prone. In general, a function that differs in the lifetime semantics of the values it returns is not good.

File: cpp/src/lists/contains.cu (hunk 15 above)
Comment (mythrocks) [reply cpp/src/lists/contains.cu]: This is interesting. I wonder why this error did not manifest in the old code.

File: cpp/src/lists/contains.cu [hunk 18]
```diff
@@ -44,296 +41,241 @@
...
+  template <bool find_first>
+  static __device__ size_type search_list(list_device_view const& list, Type const& search_key)
   {
-    auto const list_begin = find_begin<ElementType, find_option>(list);
-    auto const list_end   = find_end<ElementType, find_option>(list);
-    auto const find_iter  = thrust::find_if(
-      thrust::seq, list_begin, list_end, [search_key] __device__(auto element_and_validity) {
-        auto [element, element_is_valid] = element_and_validity;
-        return element_is_valid && cudf::equality_compare(element, search_key);
+    auto const [begin, end] = element_index_pair_iter<find_first>(list.size());
```
Comment (jrhemstad) [inline cpp/src/lists/contains.cu:111]: Is this function going to be used elsewhere? Otherwise I think it would be much easier to read as an IILE

File: cpp/src/lists/contains.cu [hunk 19]
```diff
@@ -44,296 +41,241 @@
...
+struct search_lists_fn {
+  cudf::detail::lists_column_device_view const lists;
+  SearchKeyIter const keys_iter;
+  duplicate_find_option const find_option;
 
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return find_iter - begin;  // Distance of find_position from begin.
-  } else {
-    return end - find_iter - 1;  // Distance of find_position from end.
+  __device__ thrust::pair<size_type, bool> operator()(size_type const idx) const
```
Comment (jrhemstad) [inline cpp/src/lists/contains.cu:90]: Using a `thrust::optional` would be preferred to a `pair<T, bool>`

File: cpp/src/lists/contains.cu [hunk 20]
```diff
@@ -44,296 +41,241 @@
...
-/**
- * @brief __device__ functor to search for a key in a `list_device_view`.
- */
-template <duplicate_find_option find_option>
-struct finder {
-  template <typename ElementType>
-  __device__ size_type operator()(list_device_view const& list, ElementType const& search_key) const
+ private:
+  template <bool find_first>
+  static __device__ size_type search_list(list_device_view const& list, Type const& search_key)
```
Comment (bdice) [inline cpp/src/lists/contains.cu:109]: I'd prefer to see  `search_list` as a free function like `element_index_pair_iter`, since it doesn't use any of the functor's state (`lists`, `keys_iter`, `find_option`).

File: cpp/src/lists/contains.cu (hunk 17 above)
Comment (mythrocks) [reply cpp/src/lists/contains.cu:129]: Hmm. In fairness, this predates the current change. (i.e. The error is mine.)

The aim of this was to make the lookup code common between the `column_view` and `scalar` paths. There might be a safer way to do this, but it's eluded me thus far.

File: cpp/src/lists/contains.cu [hunk 21]
```diff
@@ -44,296 +41,241 @@
...
-                            cudf::mutable_column_device_view ret_positions,
-                            cudf::mutable_column_device_view ret_validity,
-                            rmm::cuda_stream_view stream) const
+/**
+ * @brief Dispatch functor to search for key element(s) in the corresponding rows of a lists column.
+ */
+struct dispatch_index_of {
+  template <typename Type,
+            typename SearchKeyType,
+            CUDF_ENABLE_IF(!is_supported_non_nested_type<Type>())>
```
Comment (bdice) [inline cpp/src/lists/contains.cu:148]: I'd prefer to see the valid case defined first, and the failure case defined second.

File: cpp/src/lists/contains.cu [hunk 22]
```diff
@@ -44,296 +41,241 @@
...
   }
 
-  template <typename ElementType, typename SearchKeyType>
-  std::enable_if_t<is_supported<ElementType>::value, std::unique_ptr<column>> operator()(
-    cudf::lists_column_view const& lists,
-    SearchKeyType const& search_key,
-    duplicate_find_option find_option,
-    rmm::cuda_stream_view stream,
-    rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource()) const
+  template <typename Type,
```
Comment (bdice) [inline cpp/src/lists/contains.cu:159]: I recommend `typename ElementType` or `typename Element` over `typename Type`.

File: cpp/src/lists/contains.cu [hunk 23]
```diff
@@ -44,296 +41,241 @@
...
- * the search key was found.
+ * @brief Converts key-positions vector (from `index_of()`) to a BOOL8 vector, indicating if
+ *        the search key(s) were found.
  */
 std::unique_ptr<column> to_contains(std::unique_ptr<column>&& key_positions,
                                     rmm::cuda_stream_view stream,
                                     rmm::mr::device_memory_resource* mr)
 {
   CUDF_EXPECTS(key_positions->type().id() == type_id::INT32,
                "Expected input column of type INT32.");
```
Comment (bdice) [inline cpp/src/lists/contains.cu:222]: We actually interpret this column as `size_type` so the `CUDF_EXPECTS` should mirror that. We should avoid hard-coding the definition that `size_type` is INT32.
```suggestion
  CUDF_EXPECTS(key_positions->type().id() == type_to_id<size_type>,
               "Expected input column of type cudf::size_type.");
```

File: cpp/src/lists/contains.cu (hunk 15 above)
Comment (bdice) [reply cpp/src/lists/contains.cu]: > Could this be due to anonymous space?

The anonymous namespace sounds like a likely culprit to me.

Comment (karthikeyann) [reply cpp/src/lists/contains.cu]: I tried it in Godbolt too, even anonymous namespace. could not replicate it in 11.5, 11.2.

Comment (bdice) [reply cpp/src/lists/contains.cu]: Thanks for that @PointKernel. I didn't realize that guidance applied to variables as well as functions, but I agree it's reasonable to prefer the explicit nature of `__device__`.

File: cpp/src/lists/contains.cu (hunk 18 above)
Comment (ttnghia) [reply cpp/src/lists/contains.cu:111]: This will be reused in the follow-up PR when nested types implementation is added.

File: cpp/src/lists/contains.cu (hunk 19 above)
Comment (ttnghia) [reply cpp/src/lists/contains.cu:90]: The returning pair will be written into a zip iterator of `<size_type, bool>`. Can we use `thrust::optional` here for such output?

File: cpp/src/lists/contains.cu [hunk 24]
```diff
@@ -44,296 +41,241 @@
...
-  return &search_key;
-}
+/**
+ * @brief A sentinel value used for marking that a given key has not been found in the search list.
+ */
+auto constexpr __device__ NOT_FOUND_SENTINEL = size_type{-1};
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_begin(list_device_view const& list)
+template <typename Type>
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/lists/contains.cu:53]: I changed this because `ElementType` seems too long and that expands the code lines. Then `Element` seems to be used more often. I'll change to it.

File: cpp/src/lists/contains.cu (hunk 20 above)
Comment (ttnghia) [reply cpp/src/lists/contains.cu:109]: The function `search_list` here is used exclusively by the `operator()` above it. When nested types are added, there will be another pair of `search_lists_nested_fn` struct in which its `operator()` calls to `search_list_nested` function. So, I would prefer to bind the `search_list*` functions into their callers.

File: cpp/src/lists/contains.cu [hunk 25]
```diff
@@ -44,296 +41,231 @@
...
-                                      stream);
-
-    auto [null_mask, num_nulls] = construct_null_mask(lists, result_validity->view(), stream, mr);
-    result_positions->set_null_mask(std::move(null_mask), num_nulls);
-    return result_positions;
+    auto const lists_cdv_ptr = column_device_view::create(lists.parent(), stream);
+    auto const lists_cdv     = cudf::detail::lists_column_device_view{*lists_cdv_ptr};
+
+    auto out_positions = make_numeric_column(
+      data_type{type_to_id<size_type>()}, lists.size(), cudf::mask_state::UNALLOCATED, stream, mr);
```
Comment (jrhemstad) [inline cpp/src/lists/contains.cu:164]: (For a future issue/PR) We should add a `column` constructor to be constructible from a `device_uvector&&` so for outputs like this we can just more conveniently construct a `device_uvector<size_type>` and then construct the column when returning the output:
```
rmm::device_uvector<size_type> out_positions{...};
...
return make_unique<column>{std::move(out_positions)};
```
```

File: cpp/src/lists/contains.cu (hunk 19 above)
Comment (jrhemstad) [reply cpp/src/lists/contains.cu:90]: So after looking further at how this is being used, I think this is a rare instance where the use of a `pair<T,bool>` makes sense because of the fact that we're outputting to a zip iterator over two arrays. 

The alternative would be to return an `optional<T>` and add a further transform output iterator to convert an `optional<T>` to a `pair<T,bool>`, but that seems unnecessary.

File: cpp/src/lists/contains.cu [hunk 26]
```diff
@@ -44,296 +41,231 @@
...
+  duplicate_find_option const find_option;
 
-template <duplicate_find_option find_option, typename Iterator>
-size_type __device__ distance([[maybe_unused]] Iterator begin, Iterator end, Iterator find_iter)
-{
-  if (find_iter == end) {
-    return absent_index;  // Not found.
+  __device__ thrust::pair<size_type, bool> operator()(size_type const idx) const
+  {
+    auto const list = list_device_view{lists, idx};
```
Comment (jrhemstad) [inline cpp/src/lists/contains.cu:92]: This functor takes the `idx` just to construct the `list_device_view`. Does `lists_column_device_view` have an iterator that just constructs a `list_device_view` for each of the elements in the column? 

Then this functor could accept a `list_device_view` as the input and the algorithm could be a `transform` instead of a `tabulate` (which would be more natural).

Comment (ttnghia) [reply cpp/src/lists/contains.cu:92]: `idx` is also used to retrieve key, as below (line 97).

File: cpp/src/lists/contains.cu (hunk 25 above)
Comment (bdice) [reply cpp/src/lists/contains.cu:164]: Issue filed: https://github.com/rapidsai/cudf/issues/11115

File: cpp/src/lists/contains.cu [hunk 27]
```diff
@@ -32,308 +30,256 @@
...
-  } else {
-    return thrust::make_reverse_iterator(list.pair_rep_begin<ElementType>());
-  }
-}
+  template <typename Element, CUDF_ENABLE_IF(is_supported_non_nested_type<Element>())>
+  __device__ size_type operator()(list_device_view list, thrust::optional<Element> key_opt) const
+  {
+    // A null list never contains any key, even null key.
+    // In addition, a null list will result in a null output row.
+    if (list.is_null()) { return NULL_SENTINEL; }
```
Comment (jrhemstad) [inline cpp/src/lists/contains.cu:79]: Why have two separate `if` statements that return the same thing?

File: cpp/src/lists/contains.cu [hunk 28]
```diff
@@ -32,308 +30,256 @@
...
-    return absent_index;  // Not found.
-  }
+    // A null key will also result in a null output row.
+    if (!key_opt) { return NULL_SENTINEL; }
 
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return find_iter - begin;  // Distance of find_position from begin.
-  } else {
-    return end - find_iter - 1;  // Distance of find_position from end.
+    auto const& key = key_opt.value();
```
[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/src/lists/contains.cu:84]: Also, this seems like an unnecessary local variable instead of just doing:
```
  return find_option == duplicate_find_option::FIND_FIRST
             ? search_list<true, Element>(list, *key_opt)
             : search_list<false, Element>(list, *key_opt);
```

File: cpp/src/lists/contains.cu [hunk 29]
```diff
@@ -32,308 +30,256 @@
...
-    auto const list_end   = find_end<ElementType, find_option>(list);
-    auto const find_iter  = thrust::find_if(
-      thrust::seq, list_begin, list_end, [search_key] __device__(auto element_and_validity) {
-        auto [element, element_is_valid] = element_and_validity;
-        return element_is_valid && cudf::equality_compare(element, search_key);
-      });
-    return distance<find_option>(list_begin, list_end, find_iter);
-  };
-};
+    cudf_assert(false && "Nested types are not yet supported");
```
Comment (jrhemstad) [inline cpp/src/lists/contains.cu:94]: `CUDF_UNREACHABLE`.

File: cpp/src/lists/contains.cu [hunk 30]
```diff
@@ -32,308 +30,256 @@
...
-      cudf::is_numeric<ElementType>() || cudf::is_chrono<ElementType>() ||
-      cudf::is_fixed_point<ElementType>() || std::is_same_v<ElementType, cudf::string_view>;
-  };
-
-  template <typename ElementType, typename... Args>
-  std::enable_if_t<!is_supported<ElementType>::value, std::unique_ptr<column>> operator()(
-    Args&&...) const
+ private:
+  template <bool forward, typename Element, CUDF_ENABLE_IF(is_supported_non_nested_type<Element>())>
+  static __device__ size_type search_list(list_device_view const& list, Element const& search_key)
```
Comment (jrhemstad) [inline cpp/src/lists/contains.cu:99]: I strongly recommend against passing thread local objects by reference in device code as it can impact local memory usage if the compiler decides that an ODR-use requires it to put it in memory whose address can be taken. 
```suggestion
  static __device__ size_type search_list(list_device_view const list, Element const search_key)
```

Comment (ttnghia) [reply cpp/src/lists/contains.cu:99]: I was once recommended to pass user-defined objects by ref so the amount of mem copy can be minimized.

**Review by bdice — APPROVED** [review]
Nice. This has improved a lot with review.

**Comment by ttnghia** [comment]
Thanks all for reviewing. Yeah, the implementation has been improved significantly with your help :heart:

File: cpp/src/lists/contains.cu [hunk 31]
```diff
@@ -33,308 +31,251 @@
...
+    return find_option == duplicate_find_option::FIND_FIRST
+             ? search_list<true, Element>(list, *key_opt)
+             : search_list<false, Element>(list, *key_opt);
   }
 
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return find_iter - begin;  // Distance of find_position from begin.
-  } else {
-    return end - find_iter - 1;  // Distance of find_position from end.
+  template <typename Element, CUDF_ENABLE_IF(cudf::is_nested<Element>())>
```
Comment (karthikeyann) [inline cpp/src/lists/contains.cu:86]: Shouldn't the condition for enable_if be negation of other one?
```suggestion
  template <typename Element, CUDF_ENABLE_IF(!is_supported_non_nested_type<Element>())>
```

File: cpp/src/lists/contains.cu [hunk 32]
```diff
@@ -33,308 +31,251 @@
...
 
-/**
- * @brief __device__ functor to search for a key in a `list_device_view`.
- */
-template <duplicate_find_option find_option>
-struct finder {
-  template <typename ElementType>
-  __device__ size_type operator()(list_device_view const& list, ElementType const& search_key) const
+ private:
+  template <bool forward, typename Element, CUDF_ENABLE_IF(is_supported_non_nested_type<Element>())>
```
Comment (karthikeyann) [inline cpp/src/lists/contains.cu:93]: SFINAE may not be required here since the calling code has it. But it is fine to have it too.

**Review by karthikeyann — APPROVED** [review]
Minor comments. Rest look good. 👍

**Comment by ttnghia** [comment]
@gpucibot merge
